File: card-api/src/main/java/com/jean202/cardmizer/api/normalization/MerchantNormalizationRules.java
package com.jean202.cardmizer.api.normalization;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

public record MerchantNormalizationRules(
        Map<String, String> categoryAliases,
        Map<String, String> tagAliases,
        Set<String> simplePayTags,
        Set<String> onlineCategories,
        Set<String> offlineCategories,
        Set<String> subscriptionCategories,
        Set<String> digitalContentCategories,
        List<MerchantNormalizationRule> merchantRules
) {
    public MerchantNormalizationRules {
        categoryAliases = Map.copyOf(Objects.requireNonNull(categoryAliases, "categoryAliases must not be null"));
        tagAliases = Map.copyOf(Objects.requireNonNull(tagAliases, "tagAliases must not be null"));
        simplePayTags = Set.copyOf(Objects.requireNonNull(simplePayTags, "simplePayTags must not be null"));
        onlineCategories = Set.copyOf(Objects.requireNonNull(onlineCategories, "onlineCategories must not be null"));
        offlineCategories = Set.copyOf(Objects.requireNonNull(offlineCategories, "offlineCategories must not be null"));
        subscriptionCategories = Set.copyOf(
                Objects.requireNonNull(subscriptionCategories, "subscriptionCategories must not be null")
        );
        digitalContentCategories = Set.copyOf(
                Objects.requireNonNull(digitalContentCategories, "digitalContentCategories must not be null")
        );
        merchantRules = List.copyOf(Objects.requireNonNull(merchantRules, "merchantRules must not be null"));
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/normalization/MerchantNormalizationRules.kt
package com.jean202.cardmizer.api.normalization

class MerchantNormalizationRules(
    categoryAliases: Map<String, String>,
    tagAliases: Map<String, String>,
    simplePayTags: Set<String>,
    onlineCategories: Set<String>,
    offlineCategories: Set<String>,
    subscriptionCategories: Set<String>,
    digitalContentCategories: Set<String>,
    merchantRules: List<MerchantNormalizationRule>,
) {
    val categoryAliases: Map<String, String> = categoryAliases.toMap()
    val tagAliases: Map<String, String> = tagAliases.toMap()
    val simplePayTags: Set<String> = simplePayTags.toSet()
    val onlineCategories: Set<String> = onlineCategories.toSet()
    val offlineCategories: Set<String> = offlineCategories.toSet()
    val subscriptionCategories: Set<String> = subscriptionCategories.toSet()
    val digitalContentCategories: Set<String> = digitalContentCategories.toSet()
    val merchantRules: List<MerchantNormalizationRule> = merchantRules.toList()
}
